File: _58_SpiralPrimes.java
import java.util.*;

public class _58_SpiralPrimes{
    
    public static boolean isPrime(int n){
        for (int i = 2; i*i <= n; i++){
            if (n % i == 0){
                return false;
            }
        }
        
        return true;
    }
    
    public static void main(String[] args){
        
        // Corners of 3x3 square: 3, 5, 7, 9
        // Corners of 5x5 square: 13, 17, 21, 25
        // Corners of 7x7 square: 31, 37, 43, 49
        
        // Corners of nxn square: n*n, n*n - n + 1, n*n - 2*n + 2, n*n - 3*n + 3
        
        // Numbers in diagonals of 3x3 square: 5
        // Numbers in diagonals of 5x5 square: 9
        // Numbers in diagonals of 7x7 square: 13
        
        int length = 3;
        int numPrimes = 3; // 3, 5, 7
        int corner = 9;
        
        while (numPrimes * 10 >= (2*length - 1)){
            length += 2;
            
            // Check the primality of the four corners.
            for (int i = 0; i <= 3; i++){
                corner += length - 1;
                
                if (isPrime(corner)){
                    numPrimes++;
                }
            }
            // System.out.println(numPrimes + "/" + (2*length - 1) + " = " + numPrimes / (2.0*length - 1));
        }
        
        System.out.println("The ratio of primes along both diagonals first falls below 10% when the side length is " + length); // 26241
    }
}

File: cpp/58_spiral_primes.cpp
#include <iostream>

bool is_prime(int n){
    for (int i = 2; i * i <= n; i++){
        if (n % i == 0) return false;
    }
    return true;
}

int spiral_primes(){
    // 3x3: 3, 5, 7, 9
    // 5x5: 13, 17, 21, 25
    // 7x7: 31, 37, 43, 49
    // nxn: n^2 - 3n + 3, n^2 - 2n + 2, n^2 - n + 1, n^2
    
    // 2*n - 1 numbers in diagonals for square of side length n
    
    // primes / diag_nums >= 0.1
    // 10 * primes / diag_nums >= 1
    // 10 * primes >= diag_nums
    // 10 * primes >= 2*n - 1
    
    // square spiral with side length 3 has 
    // 3 primes (3, 5, 7) out of 5 numbers on diagonals (1, 3, 5, 7, 9)
    int n = 3, primes = 3;
    
    while (10 * primes >= 2 * n - 1){
        n += 2;
        for (int c = 0; c < 4; c++){
            if (is_prime(n*n - c*(n - 1))) primes++;
        }
    }
    return n;
}

int main(){
    std::cout << spiral_primes() << "\n";
}
